### PR TITLE
Fix `ossl_x509v3_cache_extensions()`: `EXFLAG_NO_FINGERPRINT` is not an error

### PR DESCRIPTION
**Pull request by DDvO**
Upon closer inspection, it turns out that the root cause of the hickup discussed for #16036 is in `ossl_x509v3_cache_extensions()`: 
the case `EXFLAG_NO_FINGERPRINT` simply should not be regarded as an error.

This PR fixes the problem and on this occasion cleans up the error handling of that function.
The fix allows reverting the recent workaround in commit 6bfd3e51c04faa97ed98f38e35bd9bb5294b9070 on `cmp_ctx_test.c` regarding `X509_new()`.

- [x] tests are added or updated

### CODE REVIEW

**Comment by levitte** [comment]
Have you looked in history why this was made an error?

What happens with OSSL_CMP_CTX_set1_srvCert()? It shouldn't accept an invalid certificate, should it?

**Comment by levitte** [comment]
There's been a few changes over time...  at some point, a certificate that was impossible to make a fingerprint from was regarded as invalid.  Why should that no longer be the case?  I find it deeply troublesome that, by being sloppy with this, functions like `OSSL_CMP_CTX_set1_srvCert()` will potentially accept invalid certificates without even noticing!

**Comment by DDvO** [comment]
> Have you looked in history why this was made an error?

I had introduced this myself in commit 1e41dadfa7b9f792ed0f4714a3d3d36f070cf30e
and meanwhile I see it was wrong.

> What happens with OSSL_CMP_CTX_set1_srvCert()? It shouldn't accept an invalid certificate, should it?

Actually the term 'invalid certificate' is very vague, and I should not have used that therm in `OSSL_CMP_CTX_set1_srvCert()` and friends.

The function `X509v3_cache_extensions()` has grown wild over the years, and at least its name should be changed because it constantly causes confusion. It does not do full validity check on the given cert, but just caches some stuff and doing so needs to perform some very basic well-formedness checks.

> I find it deeply troublesome that, by being sloppy with this, functions like `OSSL_CMP_CTX_set1_srvCert()` will potentially accept invalid certificates without even noticing!

No need to worry here - to admit, my definition
```
#define X509_invalid(cert) (!ossl_x509v3_cache_extensions(cert))
```
uses too strong words - what is more adequate is the comment soon thereafter:
```
    /* prevent misleading error later on malformed cert or provider issue */ \
```
So the function is just called for improving diagnostics.
Even if this was omitted, no insecurity is introduced because the cert is anyway verified at a later stage where needed.

**Comment by mattcaswell** [comment]
OTC: This change is not for 3.0. It can be considered post 3.0.

**Comment by DDvO** [comment]
Thanks @t8m for resuming the handling of this rather old PR - I was about to ask for that, but then saw your approval :)
CI did pass meanwhile.

So now asking Committers for the 2nd review.

**Comment by DDvO** [comment]
Rebased to fix merge conflict.
Asking again @openssl/committers for2nd review.

File: crypto/x509/v3_purp.c
Comment (beldmit) [inline crypto/x509/v3_purp.c:425]: Shouldn't we also remove call to `EVP_sha1` here?

File: crypto/x509/v3_purp.c
Comment (beldmit) [inline crypto/x509/v3_purp.c:632]: Why is this check moved from lock?

File: crypto/x509/v3_purp.c
Comment (DDvO) [reply crypto/x509/v3_purp.c:425]: Maybe, but there are also many other places where `EVP_sha1()` is still used,
and if this should be changed, it is not in scope of this PR.

File: crypto/x509/v3_purp.c
Comment (DDvO) [reply crypto/x509/v3_purp.c:632]: The lock is just in this function for consistently updating the cert structure.
Reading a single field, e.g., `ex_flags`, does not need such a protection. In  `crypto/x509/` you can find many such instances.

Comment (beldmit) [reply crypto/x509/v3_purp.c:632]: I have some suspicions, but prefer not to change the behavior - e.g. to avoid thread races.

Comment (DDvO) [reply crypto/x509/v3_purp.c:633]: So moved the check under the lock again.

**Review by beldmit — APPROVED** [review]
LGTM if tests passed

**Comment by beldmit** [comment]
@t8m could you please reconfirm?

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Is this for master or master and 3.0?

**Comment by t8m** [comment]
Master only as this kind of refactoring is IMO not a bug fix.

**Comment by DDvO** [comment]
Merged - thanks @levitte, @t8m and @beldmit

**Comment by DDvO** [comment]
> Master only as this kind of refactoring is IMO not a bug fix.

Well, to me, removing a wrongly thrown error is a bug fix, though a minor one.
Do you see a problem backporting this?

**Comment by t8m** [comment]
I wouldn't recommend backporting this. It is really border-line on whether this is a bug fix or not.

**Comment by DDvO** [comment]
All right, so closing.

**Pull request closed without merging**
